### ISSUE TITLE
XEP Push Notifications implementation

### ISSUE THREAD

**Issue by royneary** [legastero/customxeps#2]
I'm trying to implement mod_push for ejabberd according to your draft. Some ideas and questions have come up and if you have any spare time I'd be very interested in your thoughts.

For the scope of this issue I assume there's a client named chatninja and an xmpp server with hostname chatninja.org.
- Did you consider a client backend service could be the same instance as the user's server (the push service needs a dedicated jid of course)? The ejabberd-specific advantage of this would be a simple integration with ejabberd's clustering feature.
  In this case no push notification would have to be crafted but instead the contents could be passed directly to the internal proprietary push notification crafter function. If you're ok with this kind of setup, I think it should be mentioned in the "Delivering a Push Notification" section. I'm not sure what this means for the feature 'urn:xmpp:push:target' you proposed in section "Where to Deliver Notifications", as there would be no stream and thus no feature negotiation.
- I suppose you planned to suggest the push service could be an xmpp component as defined in XEP-0114 and maybe also XEP-0225. If I'm not mistaken XEP-0114 allows the component name to be a jid, e.g. 'chatninja.org/push-apns-chatninja' whereas XEP-0225 only allows components to register with a dedicated hostname, e.g. 'push-apns-chatninja.chatninja.org'. I'm bringing this up, because a service with a jid means less administration effort (no DNS A or SRV records have to be added for service to be reachable from other xmpp servers), so maybe it should be proposed somewhere in the standard (with a hint saying XEP-0225 does not allow jids).
- concerning Service Discovery I think an xmpp server should advertise push services it's controlling. I'm thinking of two different ways, one only using XEP-0030 and the other using XEP-0128. I find the second approach somewhat cleaner, XEP-0128 has not been implemented very often though.
  In both cases an disco#info query should return a push node which can be queried to get the service jids.

``` xml
C:
<iq type='get'
    from='alice@chatninja.org'
    to='chatninja.org'
    id='disco1'>
  <query xmlns='http://jabber.org/protocol/disco#items'/>
</iq>

S:
<iq type='result'
    from='chatninja.org'
    to='alice@chatninja.org'
    id='disco1'>
  <query xmlns='http://jabber.org/protocol/disco#items'>
        <item jid='chatninja.org'
              node='urn:xmpp:push:0' />
  </query>
</iq>
```

advertising push services using XEP-0030:

``` xml
C:
<iq type='get'
    from='alice@chatninja.org'
    to='chatninja.org'
    id='disco2'>
  <query xmlns='http://jabber.org/protocol/disco#items'
         node='urn:xmpp:push:0' />
</iq>

S:
<iq type='result'
    from='chatninja.org'
    id='disco2'>
  <query xmlns='http://jabber.org/protocol/disco#items'
         node='urn:xmpp:push:0'>
        <item jid='chatninja.org/push-up'
              node='urn:xmpp:push:0#up'
              name='Ubuntu Push backend'/>
        <item jid='chatninja.org/push-apns-chatninja'
              node='urn:xmpp:push:0#chatninja@apns'
              name='Apple Push Notification Service backend for client app chatninja'/>
        <item jid='chatninja.org/push-gcm-chatninja'
              node='urn:xmpp:push:0#chatninja@gcm'
              name='Google Cloud Messaging backend for client app chatninja'/>
        <item jid='chatninja.org/push-wns-chatninja'
              node='urn:xmpp:push:0#chatninja@wns'
              name='Windows Push Notification Services backend for client app chatninja'/>
  </query>
</iq>
```

advertising push services using XEP-0128:

``` xml
C:
<iq type='get'
    from='alice@chatninja.org'
    to='chatninja.org'
    id='disco1'>
  <query xmlns='http://jabber.org/protocol/disco#info'
         node='urn:xmpp:push:0' />
</iq>

S:
<iq type='result'
    from='chatninja.org'
    id='disco1'>
  <query xmlns='http://jabber.org/protocol/disco#info'
         node='urn:xmpp:push:0'>
    <identity category='push'
              type='up'
              name='Ubuntu Push backend'/>
    <identity category='push'
              type='chatninja@apns'
              name='Apple Push Notification Service backend for client app chatninja'/>
    <identity category='push'
              type='chatninja@gcm'
              name='Google Cloud Messaging backend for client app chatninja'/>
    <identity category='push'
              type='chatninja@wns'
              name='Windows Push Notification Services backend for client app chatninja'/>
    <x xmlns='jabber:x:data' type='result'>
      <field var='FORM_TYPE' type='hidden'>
        <value>urn:xmpp:push:0#backends</value>
      </field>
      <field var='push#up'>
        <value>chatninja.org/push-up</value>
      </field>
      <field var='push#chatninja@apns'>
        <value>chatninja.org/push-apns-chatninja</value>
      </field>
      <field var='push#chatninja@gcm'>
        <value>chatninja.org/push-gcm-chatninja</value>
      </field>
      <field var='push#chatninja@wns'>
        <value>chatninja.org/push-wns-chatninja</value>
      </field>
    </x>
  </query>
</iq>
```
- Did you think about requiring the use of XEP-0352? Otherwise the standard would need to define filter rules for incoming stanzas (assuming sending push notifications on incoming chat state notifications or presence updates is not desired).
- My implementation aims to make use of XEP-0198: If a client lost connection and timeout has not expired, an incoming message shall trigger a push notification to wake up the client so that it can resume the stream and the message can be delivered. Was the 'include-bodies' option meant for this purpose? Does the 'include-bodies' option being set to '1' imply that the message will be exclusively delivered by push notification? What if timeout is expired or if XEP-0198 stream resumption is not used? Then, if 'include-bodies' is set to '0' offline messages are needed for this to make any sense, right? I wouldn't complain if the standard made XEP-0198 stream resumption mandatory because it would reduce complexity but I don't know if xep-dependencies are good.
- If XEP-0198 stream resumption is used, how about the server sending a push notification to the client before the timeout expires? If the client resumes the stream the timout would be reset and thus streams would be kept in a defined state even if a client is used quite seldomly.

cheers,
Christian

**Comment by devurandom** [comment]
@royneary Did you progress with this?

**Comment by royneary** [comment]
Yes, I'm currently implementing the new version of the XEP (http://xmpp.org/extensions/inbox/push.html). It's not on github yet, but it'll arrive soon. There's a discussion on the XMPP standards mailing list about it, too.

**Comment by devurandom** [comment]
Maybe [XEP-0357](http://www.xmpp.org/extensions/xep-0357.html) is also interesting to @iNPUTmice from [siacs/Conversations](https://github.com/siacs/Conversations).

**Comment by devurandom** [comment]
Maybe @iNPUTmice wants to coordinate with you and add client-side support for [XEP-0357](http://www.xmpp.org/extensions/xep-0357.html) to [siacs/Conversations](https://github.com/siacs/Conversations)?

**Comment by royneary** [comment]
I'm using conversations and I like it. I'm not sure if they want to use push notifications though. In their design principles they write: "Do not require a Google Account or specifically Google Cloud Messaging (GCM)".

GCM is on my list of supported services. So are Ubuntu touch (already implemented), APNS and WNS. I'm sure the conversation devs read the standards mailing list, so if they're interested they can join the discussion.

**Comment by chrisballinger** [comment]
I'm glad to see other people working on this problem! I'm curious how this will integrate with my own [privacy-focused push server](https://github.com/ChatSecure/ChatSecure-Push-Server/tree/master/docs/v3) designs.

**Comment by royneary** [comment]
I opened a new issue (this place is a little hidden):
https://github.com/ChatSecure/ChatSecure-Push-Server/issues/4
